readme.md
![snake.js](./assets/snake.gif)

## snake.js
snake.js is a multiplayer game built on Node.JS and WebGL.

[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)

## Controls

- Move [W, A, S, D]
- Boost [Space]
- Disable/Enable animations [V]

## Install
```sh
npm install
npm start
```

## Config

**config.json** can be found in root directory.

changelog.md
# Changelog

## [1.0.0] - 2018-11-02
### Added
- Initial release.

## [1.1.0] - 2018-14-02
### Removed
- Performance metrics.

## [1.2.0] - 2018-14-02
### Fixed
- Food can be spawned on snake model
- Snake can be spawned on another snake


## [1.2.1] - 2020-19-09
### Changed
- Removed odometer
- Updated README.md

index.js
"use strict";

var http = require("http");
var express = require("express");
var app = express();

const WebSocket = require("ws");

var interval;
const config = require("./config.json");

app.use("/", express.static(__dirname + "/public_html"));

var server = http.Server(app);

const wss = new WebSocket.Server({
  server: server,
  perMessageDeflate: true,
});

server.listen(config.port, function () {
  console.log("Server is listening on 127.0.0.1:" + config.port);
});

var colors = [0, 1, 2, 3, 4, 5, 6];

var counter = 1;

var food = [];

var size = {
  height: 600,
  width: 720,
};

function addFood() {
  if (food.length >= 24) {
    return;
  }

  var x = randomNumber(1, size.width / 10 - 2);
  var y = randomNumber(1, size.height / 10 - 2);

  var found = false;

  for (var i = 0; i < food.length; i++) {
    if (food[i].x === x && food[i].y === y) {
      found = true;
      break;
    }
  }

  wss.clients.forEach(function (socket) {
    if (found) return;
    if (!socket.data.snake) return;
    socket.data.snake.forEach(function (position) {
      if (found) return;
      if (position.x === x && position.y === y) {
        found = true;
      }
    });
  });

  if (found) {
    addFood();
  } else {
    eat = true;
    food.push({
      x: x,
      y: y,
    });
  }
}

addFood();

wss.on("close", function (ws) {});

wss.on("error", function (error) {
  console.log(error);
});

wss.on("connection", function connection(ws) {
  ws.on("error", function () {});

  ws.data = {
    id: counter++,
    color: colors.shift(),
    frame: 0,
  };

  spawn(ws);

  ws.on("close", function () {
    colors.push(ws.data.color);
  });

  ws.on("message", function incoming(message) {
    try {
      var data = JSON.parse(message);
    } catch (error) {
      return;
    }

    switch (data.type) {
      case "direction":
        ws.data.direction = data.direction;

        break;

      case "latency":
        if (ws.readyState === WebSocket.OPEN) {
          ws.send(
            JSON.stringify({
              t: 1,
              d: data.time,
            })
          );
        }

        break;

      case 70:
        addFood();
        break;

      case "boost":
        if (ws.data.snake.length > 3) {
          ws.data.snake.splice(0, 1);
          tick(ws);
        }

        break;
    }
  });
});

interval = setInterval(tick, 1000 / config.tick);

var eat = false;

function tick(ws) {
  var buffer = [];

  wss.clients.forEach(function (client) {
    if (ws && ws.data.id !== client.data.id) {
      buffer.push(client.data);
      return;
    }

    var last = client.data.snake[client.data.snake.length - 1];

    switch (client.data.direction) {
      case "right":
        if (client.data.last_direction === "left") {
          client.data.direction = client.data.last_direction;

          client.data.snake.push({
            x: last.x - 1,
            y: last.y,
          });
        } else {
          client.data.snake.push({
            x: last.x + 1,
            y: last.y,
          });
        }

        break;

      case "down":
        if (client.data.last_direction === "up") {
          client.data.direction = client.data.last_direction;

          client.data.snake.push({
            x: last.x,
            y: last.y - 1,
          });
        } else {
          client.data.snake.push({
            x: last.x,
            y: last.y + 1,
          });
        }

        break;

      case "up":
        if (client.data.last_direction === "down") {
          client.data.direction = client.data.last_direction;

          client.data.snake.push({
            x: last.x,
            y: last.y + 1,
          });
        } else {
          client.data.snake.push({
            x: last.x,
            y: last.y - 1,
          });
        }

        break;

      case "left":
        if (client.data.last_direction === "right") {
          client.data.direction = client.data.last_direction;

          client.data.snake.push({
            x: last.x + 1,
            y: last.y,
          });
        } else {
          client.data.snake.push({
            x: last.x - 1,
            y: last.y,
          });
        }

        break;
    }

    last = client.data.snake[client.data.snake.length - 1];

    var restore = false;

    if (!restore) {
      client.data.snake
        .slice(0, client.data.snake.length - 2)
        .forEach(function (item) {
          if (item.x === last.x && item.y === last.y) {
            restore = true;
          }
        });
    }

    if (!restore) {
      if (size.width / 10 - 1 === last.x || size.height / 10 - 1 === last.y) {
        restore = true;
      } else if (last.x === 0 || last.y === 0) {
        restore = true;
      }
    }

    if (restore) {
      spawn(client);
    } else {
      client.data.last_direction = client.data.direction;

      var remove = false;
      var tmp = [];

      food.forEach(function (item, index) {
        if (item.x === last.x && item.y === last.y) {
          remove = true;
          tmp.push(index);
        }
      });

      for (var i = tmp.length - 1; i > -1; i--) {
        food.splice(tmp[i], 1);
        eat = true;
      }

      if (!remove) {
        client.data.snake.splice(0, 1);
      } else {
        if (food.length === 0) {
          addFood();
        }
        client.data.score++;
      }
    }

    if (typeof client.data.color === "object") {
      if (client.data.color.array.length - 1 === client.data.color.current) {
        client.data.color.current = 0;
      } else {
        client.data.color.current++;
      }
    }

    buffer.push(client.data);
  });

  var collision = [];

  wss.clients.forEach(function (client) {
    var client_last = client.data.snake[client.data.snake.length - 1];

    wss.clients.forEach(function (another_client) {
      if (client.data.id !== another_client.data.id) {
        for (var i = 0; i < another_client.data.snake.length; i++) {
          if (
            another_client.data.snake[i].x === client_last.x &&
            another_client.data.snake[i].y === client_last.y
          ) {
            collision.push(client);
            break;
          }
        }
      }
    });
  });

  collision.forEach(function (item) {
    spawn(item);
  });

  var compressed = [];

  buffer.forEach(function (client) {
    var tmp = {
      x: [],
      y: [],
      d: client.last_direction ? client.last_direction : client.direction,
      i: client.id,
    };

    for (var b = 0; b < client.snake.length; b++) {
      tmp.x.push(client.snake[b].x);
      tmp.y.push(client.snake[b].y);
    }

    tmp.c = client.color;

    compressed.push(tmp);
  });

  var compressed_food = {
    x: [],
    y: [],
  };

  food.forEach(function (position) {
    compressed_food.x.push(position.x);
    compressed_food.y.push(position.y);
  });

  wss.clients.forEach(function (client) {
    if (client.readyState === WebSocket.OPEN) {
      client.send(
        JSON.stringify({
          t: 0,
          s: compressed,
          f: eat ? compressed_food : client.data.frame++ ? 0 : compressed_food,
          o: client.data.score,
          r: config.tick,
        })
      );
    } else {
      client.terminate();
    }
  });

  eat = false;
}

function spawn(client) {
  var coordinates = [];

  wss.clients.forEach(function (socket) {
    if (!socket.data.snake) return;
    socket.data.snake.forEach(function (position) {
      coordinates.push({
        x: position.x,
        y: position.y,
      });
    });
  });

  var coordinate = randomNumber(10, size.width / 10 - 15);

  var tmp = [
    {
      x: coordinate,
      y: coordinate,
    },
  ];

  for (var i = 0; i < config.default_snake_length; i++) {
    tmp.push({
      x: coordinate + i,
      y: coordinate,
    });
  }

  var collision = false;

  coordinates.forEach(function (coordinate) {
    tmp.forEach(function (position) {
      if (position.x === coordinate.x && position.y === coordinate.y) {
        collision = true;
      }
    });
  });

  if (collision) {
    spawn(client);
    return;
  }

  client.data.snake = tmp;

  client.data.direction = "right";
  client.data.last_direction = client.data.direction;
  client.data.score = 0;
}

function randomDirection() {
  switch (randomNumber(0, 3)) {
    case 0:
      return "up";
      break;
    case 1:
      return "right";
      break;
    case 2:
      return "down";
      break;
    case 3:
      return "left";
      break;
  }
}

setInterval(function () {
  if (2 > food.length) {
    addFood();
  }
}, 1000 * 30);

function heartbeat() {
  this.isAlive = true;
}

wss.on("connection", function connection(ws) {
  ws.isAlive = true;
  ws.on("pong", heartbeat);
});

const aliveInterval = setInterval(function ping() {
  wss.clients.forEach(function each(ws) {
    if (ws.isAlive === false) return ws.terminate();

    ws.isAlive = false;
    ws.ping();
  });
}, 1000 * 10);

function randomNumber(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}


public_html/js/script.js
(function () {
  var elements = [
    "canvas",
    "score",
    "latency",
    "bytes_received",
    "bytes_sent",
    "tick_rate",
    "food_counter",
    "fps_counter",
    "debug_toggle",
    "debug_inner",
  ].forEach((element) => {
    this[element] = document.getElementById(element);
  });

  var ctx = canvas.getContext("2d");

  debug_toggle.addEventListener("click", function () {
    if (debug_inner.style.display == "none") {
      debug_inner.style.display = "block";
    } else {
      debug_inner.style.display = "none";
    }
  });

  var socket = new WebSocket(
    "ws://" + location.hostname + ":" + location.port || 80
  );

  var time;

  var food = {
    x: [],
    y: [],
  };

  var session = {
    id: null,
    score: 0,
  };

  var config = {
    food_form: "square",
    animation: true,
  };

  var snakes = [];

  var parts = {};

  const colors = {
    0: "yellow",
    1: "cyan",
    2: "orange",
    3: "lightblue",
    4: "magenta",
    5: "lime",
    6: "red",
  };

  var traffic = {
    received: 0,
    sent: 0,
  };

  const size = {
    height: 600,
    width: 720,
    step: 10,
  };

  function send(data) {
    var string = JSON.stringify(data);
    traffic.sent += new TextEncoder("utf-8").encode(string).length;
    socket.send(string);
  }

  socket.onopen = function () {
    console.log("Connection established");

    time = Date.now();

    send({
      type: "latency",
      time: time,
    });
  };

  socket.onclose = function (event) {
    if (event.wasClean) {
      console.log("The connection is closed cleanly");
    } else {
      console.log("Connection failure");
    }
    console.log("Code: " + event.code + " reason: " + event.reason);
  };

  setInterval(function () {
    bytes_received.innerHTML = bytesToSize(traffic.received);
    bytes_sent.innerHTML = bytesToSize(traffic.sent);
  }, 1000 * 1);

  socket.onmessage = function (event) {
    traffic.received += new TextEncoder("utf-8").encode(event.data).length;

    var object = JSON.parse(event.data);

    switch (object.t) {
      case 0:
        if (object.f) {
          food = object.f;
        }

        snakes = object.s;

        if (!config.animation) {
          window.requestAnimationFrame(draw);
        }

        score.innerHTML = object.o;

        tick_rate.innerHTML = object.r;
        food_counter.innerHTML = food.x.length;

        break;

      case 1:
        var latencyTime = Date.now() - object.d;

        latency.innerHTML = latencyTime;

        setTimeout(function () {
          time = Date.now();
          send({
            type: "latency",
            time: time,
          });
        }, 1000);

        break;
    }
  };

  socket.onerror = function (error) {
    console.log("An error " + error.message);
  };

  function border() {
    ctx.fillStyle = "white";
    for (var i = 0; i < size.width / size.step; i++) {
      ctx.fillRect(size.step * i, 0 * size.step, size.step, size.step);
      ctx.fillRect(
        size.step * i,
        size.height - 1 * size.step,
        size.step,
        size.step
      );
    }
    for (var i = 0; i < size.height / size.step; i++) {
      ctx.fillRect(0, size.step * i, size.step, size.step);
      ctx.fillRect(
        size.width - 1 * size.step,
        i * size.step,
        size.step,
        size.step
      );
    }
  }

  function grid() {
    var fill = false;

    for (var a = 0; a < size.height; a++) {
      if (fill) {
        fill = false;
      } else {
        fill = true;
      }
      for (var i = 0; i < size.width / size.step; i++) {
        if (fill) {
          ctx.fillStyle = "black";
          ctx.fillRect(size.step * i, a * size.step, size.step, size.step);
          fill = false;
        } else {
          ctx.fillStyle = "#111";
          ctx.fillRect(size.step * i, a * size.step, size.step, size.step);
          fill = true;
        }
      }
    }
  }

  var then = 0;
  var fps = 0;

  function draw(now) {
    now *= 0.001;
    var delta = now - then;
    then = now;
    var new_fps = parseInt(1 / delta);

    ctx.fillStyle = "black";
    ctx.fillRect(0, 0, canvas.width, canvas.height);

    ctx.fillStyle = "#eee";

    for (var i = 0; i < food.x.length; i++) {
      switch (config.food_form) {
        case "square":
          ctx.fillRect(
            size.step * food.x[i],
            size.step * food.y[i],
            size.step,
            size.step
          );

          break;

        case "circle":
          ctx.beginPath();
          ctx.arc(
            size.step * food.x[i] + size.step / 2,
            size.step * food.y[i] + size.step / 2,
            size.step / 2 - 1,
            0,
            Math.PI * 2,
            true
          );
          ctx.closePath();
          ctx.fill();

          break;
      }
    }

    snakes.forEach(function (client) {
      var color = client.c;

      if (typeof color === "object") {
        color = color.array[color.current];
      }

      ctx.fillStyle = colors[color];

      var length = client.x.length - 1;

      var last = {
        x: client.x[length],
        y: client.y[length],
      };

      for (var i = 0; i < client.x.length; i++) {
        if (!config.animation) {
          ctx.fillRect(
            size.step * client.x[i],
            size.step * client.y[i],
            size.step,
            size.step
          );
        } else {
          if (i != length) {
            if (i !== 0) {
              ctx.fillRect(
                size.step * client.x[i],
                size.step * client.y[i],
                size.step,
                size.step
              );
            }
          }
        }
      }

      if (!config.animation) return;

      if (!parts[client.i]) {
        parts[client.i] = {
          _last: {
            x: last.x,
            y: last.y,
            progress: 0,
            direction: client.d,
          },
          _first: {
            x: client.x[0],
            y: client.y[0],
            animate: false,
          },
        };
      } else {
        if (parseInt(parts[client.i]._last.progress) == size.step) {
          ctx.fillRect(
            size.step * parts[client.i]._last.x,
            size.step * parts[client.i]._last.y,
            size.step,
            size.step
          );
        }

        if (
          parts[client.i]._last.x !== last.x ||
          parts[client.i]._last.y !== last.y
        ) {
          parts[client.i]._last = {
            x: last.x,
            y: last.y,
            progress: 0,
            direction: client.d,
          };
        }

        if (
          parts[client.i]._first.x !== client.x[0] ||
          parts[client.i]._first.y !== client.y[0]
        ) {
          parts[client.i]._first.progress = 0;
          parts[client.i]._first.animate = true;

          parts[client.i]._first.direction = findDirection(
            {
              x: parts[client.i]._first.x,
              y: parts[client.i]._first.y,
            },
            {
              x: client.x[0],
              y: client.y[0],
            }
          );

          parts[client.i]._first.x = client.x[0];
          parts[client.i]._first.y = client.y[0];

          var tmp_direction = findDirection(
            {
              x: parts[client.i]._first.x,
              y: parts[client.i]._first.y,
            },
            {
              x: client.x[1],
              y: client.y[1],
            }
          );

          if (parts[client.i]._first.direction != tmp_direction) {
            parts[client.i]._first.direction = tmp_direction;
          }
        }
      }

      if (parts[client.i]._first.progress !== size.step) {
        parts[client.i]._first.progress += 2;

        if (parts[client.i]._first.animate) {
          switch (parts[client.i]._first.direction) {
            case "right":
              ctx.fillRect(
                size.step * parts[client.i]._first.x +
                  parts[client.i]._first.progress,
                size.step * parts[client.i]._first.y,
                size.step,
                size.step
              );
              break;

            case "up":
              ctx.fillRect(
                size.step * parts[client.i]._first.x,
                size.step * parts[client.i]._first.y,
                size.step,
                size.step - parts[client.i]._first.progress
              );

              break;

            case "down":
              ctx.fillRect(
                size.step * parts[client.i]._first.x,
                size.step * parts[client.i]._first.y +
                  parts[client.i]._first.progress,
                size.step,
                size.step
              );

              break;

            case "left":
              ctx.fillRect(
                size.step * parts[client.i]._first.x,
                size.step * parts[client.i]._first.y,
                size.step - parts[client.i]._first.progress,
                size.step
              );

              break;
          }
        }
      } else {
        parts[client.i]._first.x = client.x[0];
        parts[client.i]._first.y = client.y[0];
      }

      if (parts[client.i]._last.progress !== size.step) {
        parts[client.i]._last.progress += 2;

        switch (parts[client.i]._last.direction) {
          case "right":
            ctx.fillRect(
              size.step * parts[client.i]._last.x,
              size.step * parts[client.i]._last.y,
              parts[client.i]._last.progress,
              size.step
            );

            break;

          case "down":
            ctx.fillRect(
              size.step * parts[client.i]._last.x,
              size.step * parts[client.i]._last.y,
              size.step,
              parts[client.i]._last.progress
            );

            break;

          case "up":
            ctx.fillRect(
              size.step * parts[client.i]._last.x,
              size.step * parts[client.i]._last.y +
                size.step -
                Math.abs(parts[client.i]._last.progress),
              size.step,
              size.step
            );

            break;

          case "left":
            ctx.fillRect(
              size.step * parts[client.i]._last.x +
                size.step -
                Math.abs(parts[client.i]._last.progress),
              size.step * parts[client.i]._last.y,
              size.step,
              size.step
            );

            break;
        }
      }

      ctx.shadowBlur = 0;
    });

    border();

    if (config.animation) {
      window.requestAnimationFrame(draw);
    }

    if (new_fps !== fps) {
      fps = new_fps;
      fps_counter.innerHTML = new_fps;
    }
  }

  window.requestAnimationFrame(draw);

  document.addEventListener("keydown", function (event) {
    switch (event.keyCode) {
      case 37:
      case 65:
        send({
          type: "direction",
          direction: "left",
        });
        break;
      case 38:
      case 87:
        send({
          type: "direction",
          direction: "up",
        });
        break;
      case 39:
      case 68:
        send({
          type: "direction",
          direction: "right",
        });
        break;
      case 40:
      case 83:
        send({
          type: "direction",
          direction: "down",
        });
        break;

      case 86: // v
        if (config.animation) {
          config.animation = false;
        } else {
          config.animation = true;
          window.requestAnimationFrame(draw);
        }

        break;
    }
  });

  var boost;

  function findDirection(first, second) {
    if (first.x === second.x) {
      if (first.y + 1 === second.y) {
        return "down";
      } else {
        return "up";
      }
    } else if (first.y === second.y) {
      if (first.x + 1 === second.x) {
        return "right";
      } else {
        return "left";
      }
    }
  }

  onkeydown = onkeyup = function (e) {
    if (e.keyCode === 32) {
      switch (e.type) {
        case "keydown":
          if (!boost) {
            boost = setInterval(function () {
              send({
                type: "boost",
              });
            }, 1000 / 12);
          }

          break;

        case "keyup":
          clearInterval(boost);
          boost = null;

          break;
      }
    }
  };

  function bytesToSize(bytes) {
    // https://stackoverflow.com/a/18650828
    var sizes = ["Bytes", "KB", "MB", "GB", "TB"];
    if (bytes === 0) return "0 Byte";
    var i = parseInt(Math.floor(Math.log(bytes) / Math.log(1024)));
    return Math.round(bytes / Math.pow(1024, i), 2) + " " + sizes[i];
  }
})();
